R/superPartition.R

```diff
@@ -0,0 +1,268 @@
+  # cs - clust_size, numeric vector of cluster sizes
+  # fd - full_data, sample x feature
+  reduce_clust <- function(cs, fd) {
+    # bookkeeping
+    largest_clust <- max(master_cluster$cluster)
+    unique_vals  <- unique(master_cluster$clust)
+    clusters     <- master_cluster$cluster
+
+    # keep looping until all clusters are under max cluster size
+    while(max(cs) > cluster_size) {
+      for (l in 1:length(cs)) {
+        # if cluster size is over max cluster size...
+        if(cs[l] > cluster_size) {
+          # cluster again
+          new_clust <- genie(t(fd[, which(clusters == unique_vals[l])]),
+                            k = ceiling(cs[l]/cluster_size),
+                            gini_threshold = 0.05)
+
+          # if genie doesn't do anything, use kmeans
+          if(length(unique(new_clust)) == 1 | min(table(new_clust)) < 50) {
+            new_clust <- kmeans(t(fd[, which(clusters == unique_vals[l])]),
+                               centers = ceiling(cs[l]/cluster_size))$cluster
+          }
+
+          # reassign cluster numbers; use <<- to rewrite data from dataframe defined outside function
+          master_cluster$cluster[which(clusters == unique_vals[l])] <<- new_clust + largest_clust
+
+          # update largest cluster number
+          largest_clust <- max(master_cluster$cluster)
+        }
+      }
+
+      # update bookkeeping variables
+      cs           <- table(master_cluster$cluster)
+      unique_vals  <- sort(unique(master_cluster$clust))
+      clusters     <- master_cluster$cluster
+    }
+  }
+
+  reduce_clust(cs = clust_size, fd = full_data)
+  message(paste0(length(unique(master_cluster$cluster)), " super clusters identified. Beginning Partition."))
+  message(paste0("Maximum cluster size: ", max(table(master_cluster$cluster))))
+
+  # set up progress bar
+  n_iter <- length(unique(master_cluster$cluster))
+  pb     <- progress_bar$new(format =
+                               "(:spin) [:bar] :percent [Elapsed time: :elapsedfull || Estimated time remaining: :eta]",
+                             total = n_iter,
+                             complete = "=",   # Completion bar character
+                             incomplete = "-", # Incomplete bar character
+                             current = ">",    # Current bar character
+                             clear = FALSE,    # If TRUE, clears the bar when finish
+                             width = 100)
+
+  ## first cluster
+  # get initial partition to build off
+  part_master <- partition(full_data[, which(master_cluster$cluster == unique(master_cluster$cluster)[1])],
+                           threshold, partitioner, tolerance, niter, x, .sep)
+
+  # update indices for each module
+  mod_rows <- grep("reduced_var_", part_master$mapping_key$variable)
+  suppressWarnings(part_master$mapping_key$indices <- full_data_col_numbers(full_data = full_data,
```

Comment:
What warnings are you suppressing here?

Reply:
I get a warning about the new argument not matching the length of the old argument (ie part_master$mapping_key$indices) even though they are both lists of the same length.

Reply:
hmm ok, I'll look into this. Generally, `suppressWarnings()` gives me itchy feelings, so I'd like to confirm it's necessary. But I'll handle that before merging.

Reply:
I bet you could also tweak the regex in the `grep()` call to only match the name exactly. I don't know which is the faster approach off the top of my head

Reply:
Ah okay that makes sense! I fixed this using `which`, since which is faster for exact matches than grep. I removed the suppressWarnings as well.

Additionally, this made me realize there may be problems if the original data has column names that contain the reduced variable naming string (parameter x). I've added a check at the beginning which requires that no column names contain x:
`if(any(grepl(x, colnames(full_data)))) stop(paste0("The prefix for new variable names, ", x, ", is contained within existing data column names. Please choose a different prefix to avoid errors."))`

I also added an additional set of tests for the errors.